draft-ietf-quic-recovery.md

```diff
@@ -519,6 +519,24 @@ bytes.
 Initial packets and Handshake packets may never be acknowledged, but they are
 removed from bytes in flight when the Initial and Handshake keys are discarded.
 
+### Speeding Up Handshake Completion
+
+When a server receives an Initial packet containing duplicate CRYPTO data,
+it can assume the client did not receive all of the server's CRYPTO data sent
+in Initial packets, or the client's estimated RTT is too small. When a
+client receives Handshake or 1-RTT packets prior to obtaining Handshake keys,
+it may assume some or all of the server's Initial packets were lost.
+
+To speed up handshake completion under these conditions, an endpoint MAY send
+a packet containing unacknowledged CRYPTO data earlier than the PTO expiry,
+subject to address validation limits. (see Section 8.1 of {{QUIC-TRANSPORT}})
+
+Peers can also use coalesced packets to ensure that each datagram elicits at least
+one acknowledgement.  For example, clients can coalesce an Initial packet
+containing PING and PADDING frames with a 0-RTT data packet and a server can
+coalesce an Initial packet containing a PING frame with one or more packets in
+its first flight.
```

Comment:
I'm not sure about this paragraph.  It is implying something about packet composition into datagrams that is inconsistent with our general approach.

Reply:
Thanks, I'm hoping to write text that's consistent with our approach(ie: one should not assume anything about other QUIC packets that happen to be in the same datagram) and yet note that implementations may want to bundle a small packet(ie: a PING) along with another packet in order to ensure one of the two receives an acknowledgement.  Feel free to suggest clearer text.

Reply:
I guess that I misread it, but here is an alternative...

An endpoint can coalesce packet from multiple packet number spaces with a probe. This allows the peer can drive loss recovery even when it is unable to process some packet types. Endpoints can include a PING frame in packets from spaces other than the one that caused the probe to be sent.

Reply:
I think what's missing is any description of _why_ an implementation might want to do that except things we don't really want to encourage.  The obvious reading is that clients are tracking whether the datagram was received, inferring that the other packets in that datagram were received but not processed (or processable?), and removing everything in that datagram from bytes-in-flight.  But it doesn't explicitly say that.  I'd either get more explicit, or not make the reference.

Reply:
My understanding is that this economizes probe packets.  Those are limited and # pn spaces > # probe packets.